UIforETW/PowerStatus.cpp

```diff
@@ -108,125 +108,179 @@ void CPowerStatusMonitor::SampleBatteryStat()
 	for (int deviceNum = 0; deviceNum < maxBatteries; deviceNum++)
 	{
 		SP_DEVICE_INTERFACE_DATA did = { sizeof(did) };
-
-		if (SetupDiEnumDeviceInterfaces(hdev,
+		const BOOL enumDeviceInterfacesResult =
+			SetupDiEnumDeviceInterfaces(hdev,
 										0,
 										&GUID_DEVCLASS_BATTERY,
 										deviceNum,
-										&did))
+										&did);
+		if (!enumDeviceInterfacesResult)
+		{
+			const DWORD lastErr = ::GetLastError( );
+			if (lastErr == ERROR_NO_MORE_ITEMS)
+			{
+				break; // Enumeration failed - perhaps we're out of items
+			}
+			continue;
+		}
+
+		DWORD bytesNeeded = 0;
+		const BOOL getDetailResult = SetupDiGetDeviceInterfaceDetail(hdev,
+										&did,
+										0,
+										0,
+										&bytesNeeded,
+										0);
+
+		if (getDetailResult)
+		{
+			throw std::logic_error("Expected SetupDiGetDeviceInterfaceDetail failure!");
+		}
+
+		if (GetLastError() != ERROR_INSUFFICIENT_BUFFER)
+		{
+			continue;
+		}
+
+
+		HLOCAL detailDataAlloc = LocalAlloc(LPTR, bytesNeeded);

```

Comment:
> A quick check confirms that LocalAlloc has been mostly obsolete since Win32 came into existence 20+ years ago

Yeah, pretty much. Raymond Chen, as always, [has some insight into this odd relic of 16-bit Windows](http://blogs.msdn.com/b/oldnewthing/archive/2004/11/01/250610.aspx).

The fact that the MSDN sample (for an API that's less than 20 years old?) uses it, makes me a bit nervous about changing it - did I miss something in the documentation?